[D1] Add e-commerce tutorial (#21125)
cloudflare/cloudflare-docs: src/content/docs/d1/tutorials/using-read-replication-for-e-com/index.mdx

---
updated: 2025-03-24
difficulty: Beginner
content_type: Tutorial
pcx_content_type: tutorial
title: Using D1 Read Replication for your e-commerce website
products:
  - D1
languages:
  - JavaScript
  - TypeScript
  - SQL
---

import {
	Render,
	Steps,
	PackageManagers,
	WranglerConfig,
	Details,
} from "~/components";

[D1 Read Replication](/d1/features/read-replication/) is a feature that allows you to replicate your D1 database to multiple regions. This is useful for your e-commerce website to reduce read latencies and improve availability. In this tutorial, you will learn how to use D1 Read Replication for your e-commerce website.

While this tutorial uses a fictional e-commerce website, the principles can be applied to any use case that requires high availability and low read latencies, for example, a news website, a social media platform, or a marketing website.

## Prerequisites

<Render file="prereqs" product="workers" />

> ## Step 1: Create a Workers project

Reviewer: I think all tutorials would benefit from having user click `Deploy to Workers` button, get the running examples, and then tutorial highlight steps as if you were building incrementally.

				document.getElementById('product-description').textContent = product[0].description;
				document.getElementById('product-price').textContent = `$${product[0].price.toFixed(2)}`;
				document.getElementById('product-stock').textContent = `${product[0].inventory} in stock`;
			}
		</script>
	</body>
</html>

```
</Details>

You now have a frontend that lists products and displays a single product. However, the frontend is not connected to the D1 database yet. If you start the development server now, you will see no products. In the next steps, you will create a D1 database and create APIs to fetch products and display them on the frontend.

## Step 3: Create a D1 database and enable read replication

Create a new D1 database by running the following command:

```sh
npx wrangler d1 create fast-commerce
````

Add the D1 bindings returned in the terminal to the `wrangler` file:

<WranglerConfig>
```toml
[[d1_databases]] 
binding = "DB"
database_name = "fast-commerce"
database_id = "YOUR_DATABASE_ID" 
```
</WranglerConfig>

Run the following command to update the `Env` interface in the `worker-congifuration.d.ts` file.

```sh
npm run cf-typegen
```

> Next, enable read replication for the D1 database by running the following command. You will need to replace `<accountid>` with your Cloudflare account ID, `<databaseId>` with the ID of the D1 database, and `$TOKEN` with your Cloudflare API token. You can learn more about it in the [Read replication documentation](/d1/features/read-replication/#enable-read-replication)

Reviewer 2: ```suggestion
Next, enable read replication for the D1 database by running the following command. You will need to replace `<accountid>` with your Cloudflare account ID, `<databaseId>` with the ID of the D1 database, and `$TOKEN` with your Cloudflare API token. You can learn more about it in the [Read replication documentation](/d1/features/read-replication/#enable-read-replication).

:::note
Read replication is only used when the Worker has been [deployed](/d1/tutorials/using-read-replication-for-e-com/#step-8-deploy-the-application). D1 does not create read replicas when you develop locally.
:::
```
The effects of RR is only visible when this has been deployed right? I think we should make a note of that here, because this tutorial does a lot of things before the deployment step.

```

Now, you can deploy the application with the following command:

```sh
npm run deploy
```

This will deploy the application to Workers and the D1 database will be replicated to the remote regions. If a user requests the application from any region, the request will be redirected to the nearest region where the database is replicated.

## Conclusion

In this tutorial, you learned how to use D1 Read Replication for your e-commerce website. You created a D1 database and enabled read replication for it. You then created an API to create and update products in the database. You also learned how to use the bookmark to get the latest data from the database.

You then created the products table in the remote database and deployed the application.

> You can use the same approach for your existing read heavy application to reduce read latencies and improve availability. If you are using an external platform to manage the content, you can connect the external platform to the D1 database so that the content is automatically updated in the database.

Reviewer 2: ```suggestion
You can use the same approach for your existing read heavy application to reduce read latencies and improve availability. If you are using an external platform to manage the content, you can connect the external platform to the D1 database, so that the content is automatically updated in the database.
```

> You can find the complete code for this tutorial in the [GitHub repository](https://github.com/harshil1712/e-com-d1-hono).

Reviewer 2: Making a note to update the link before merging.

---
updated: 2025-03-27
difficulty: Beginner
content_type: Tutorial
pcx_content_type: tutorial
title: Using D1 Read Replication for your e-commerce website
products:
  - D1
languages:
  - JavaScript
  - TypeScript
  - SQL
---

import {
	Render,
	Steps,
	PackageManagers,
	WranglerConfig,
	Details,
} from "~/components";

[D1 Read Replication](/d1/features/read-replication/) is a feature that allows you to replicate your D1 database to multiple regions. This is useful for your e-commerce website, as it reduces read latencies and improves availability. In this tutorial, you will learn how to use D1 read replication for your e-commerce website.

While this tutorial uses a fictional e-commerce website, the principles can be applied to any use-case that requires high availability and low read latencies, such as a news website, a social media platform, or a marketing website.

## Quick start

If you want to skip the steps and get started quickly, click on the below button:

[![Deploy to Cloudflare](https://deploy.workers.cloudflare.com/button)](https://deploy.workers.cloudflare.com/?url=https://github.com/harshil1712/e-com-d1-hono)

> This will create a repository in your GitHub account and deploy the application to Cloudflare Workers. It will also create and bind a D1 database, create the required tables, and add some sample data. You can visit the deployed application. However, you need to enable D1 Read Replication for your database.

Reviewer 2: ```suggestion
This will create a repository in your GitHub account and deploy the application to Cloudflare Workers. It will also create and bind a D1 database, create the required tables, and add some sample data. You can then visit the deployed application.

After deploying to Cloudflare through the button, this tutorial requires you to enable read replication for your newly created D1 database.
```
